src/main/java/org/springframework/guice/annotation/ModuleRegistryConfiguration.java

```diff
@@ -242,11 +239,10 @@ else if (e instanceof PrivateElements) {
 	protected List<Element> removeDuplicates(List<Element> elements) {
 		List<Element> duplicateElements = elements.stream()
 				.filter(e -> e instanceof Binding).map(e -> (Binding<?>) e)
-				.collect(Collectors.groupingBy(Binding::getKey)).entrySet().stream()
+				.collect(Collectors.groupingBy(ModuleRegistryConfiguration::getLinkedKeyIfRequired)).entrySet().stream()
```

Comment:
Guice 4.2 changed the implementation for how optionals are binding and broke the `BindingDeduplicationTests#verifyNoDuplicateBindingErrorWhenDedupeEnabled` test. 

Now in 4.2.0+ that optional binding appears to be an instance of `LinkedKeyBinding`, so I think we want to group these by these linked keys, since Guice uses that linked key to resolve bindings. 

```
linkedBinding.getKey()       // equals Key<Optional<SomeOptionalDependency>>
linkedBinding.getLinkedKey() // equals Key<SomeOptionalDependency>
```

Open to better suggestions if anyone has any ideas